### ISSUE TITLE
Thanks for your excellent work, We really want use yolo-world do more excellent work, can we use yolo-world trained model as base, providing embedding to fine-grained differentiation of samples for downstream tasks?

### ISSUE THREAD

**Issue by Meicsu199345** [ultralytics/ultralytics#13772]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Background
Continuing from the previous question, under the guidance of the author, we successfully applied yolo-world as a prompt, prompt like “person holding small white cigarette and smoking cigarette with mouth”, work excellent in our work.

Question
Now, we have a new question, we really want use yolo-world do more excellent work, can we use yolo-world trained model as base, providing embedding to fine-grained differentiation of samples for downstream tasks?
For example, two categories: white cigarettes and black cigarettes.
I learn yolo-world network backbone-neck-head，should i extra neck featuremap as output?
                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]              
  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]           
  7                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
  8                  -1  3   6969600  ultralytics.nn.modules.block.C2f             [640, 640, 3, True]           
  9                  -1  1   1025920  ultralytics.nn.modules.block.SPPF            [640, 640, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  3   8670410  ultralytics.nn.modules.block.C2fAttn         [1280, 640, 3, 320, 10]       
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  3   2312805  ultralytics.nn.modules.block.C2fAttn         [960, 320, 3, 160, 5]         
 16                  15  1    922240  ultralytics.nn.modules.conv.Conv             [320, 320, 3, 2]              
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  3   8465610  ultralytics.nn.modules.block.C2fAttn         [960, 640, 3, 320, 10]        
 19                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  3   8670410  ultralytics.nn.modules.block.C2fAttn         [1280, 640, 3, 320, 10]       
 22        [15, 18, 21]  1   9214102  ultralytics.nn.modules.head.WorldDetect      [80, 512, True, [320, 640, 640]]

stage0_Conv_features.npy
(80, 320, 320)
stage10_Upsample_features.npy
(640, 40, 40)
stage11_Concat_features.npy
(1280, 40, 40)
stage12_C2fAttn_features.npy
(640, 40, 40)
stage13_Upsample_features.npy
(640, 80, 80)
stage14_Concat_features.npy
Array shape:
(960, 80, 80)
filename
stage15_C2fAttn_features.npy
(320, 80, 80)
stage16_Conv_features.npy
(320, 40, 40)
stage17_Concat_features.npy
(960, 40, 40)
stage18_C2fAttn_features.npy
(640, 40, 40)
stage19_Conv_features.npy
(640, 20, 20)
filename
stage1_Conv_features.npy
Array shape:
(160, 160, 160)
stage20_Concat_features.npy
(1280, 20, 20)
stage21_C2fAttn_features.npy
(640, 20, 20)
stage2_C2f_features.npy
(160, 160, 160)
stage3_Conv_features.npy
(320, 80, 80)
stage4_C2f_features.npy
(320, 80, 80)
stage5_Conv_features.npy
(640, 40, 40)
stage6_C2f_features.npy
Array shape:
stage7_Conv_features.npy
(640, 20, 20)
stage8_C2f_features.npy
(640, 20, 20)
stage9_SPPF_features.npy
(640, 20, 20)

Related work
Previously, clip was used to extract features, and a linear regression model was used downstream to distinguish them. This eliminates the need for retraining and does not affect other labels. Added fine-grained classification capabilities;

thank you all for your answers. I wish yolo-world will get better and better.
link：
https://github.com/openai/CLIP?tab=readme-ov-file#linear-probe-evaluation
![image](https://github.com/ultralytics/ultralytics/assets/16251955/5f2a2828-5547-477b-97c3-ea25d1f68607)

`
code like
import os
import clip
import torch

import numpy as np
from sklearn.linear_model import LogisticRegression
from torch.utils.data import DataLoader
from torchvision.datasets import CIFAR100
from tqdm import tqdm

Load the model
device = "cuda" if torch.cuda.is_available() else "cpu"
model, preprocess = clip.load('ViT-B/32', device)

Load the dataset
root = os.path.expanduser("~/.cache")
train = CIFAR100(root, download=True, train=True, transform=preprocess)
test = CIFAR100(root, download=True, train=False, transform=preprocess)

def get_features(dataset):
all_features = []
all_labels = []

with torch.no_grad():
    for images, labels in tqdm(DataLoader(dataset, batch_size=100)):
        features = model.encode_image(images.to(device))

        all_features.append(features)
        all_labels.append(labels)

return torch.cat(all_features).cpu().numpy(), torch.cat(all_labels).cpu().numpy()
Calculate the image features
train_features, train_labels = get_features(train)
test_features, test_labels = get_features(test)

Perform logistic regression
classifier = LogisticRegression(random_state=0, C=0.316, max_iter=1000, verbose=1)
classifier.fit(train_features, train_labels)

Evaluate using the logistic regression classifier
predictions = classifier.predict(test_features)
accuracy = np.mean((test_labels == predictions).astype(float)) * 100.
print(f"Accuracy = {accuracy:.3f}")

**Comment by glenn-jocher** [comment]
@Meicsu199345 hello,

Thank you for your kind words and for sharing your success with YOLO-World! It's fantastic to hear that the model has been effective in your work.

Regarding your question about using YOLO-World for fine-grained differentiation of samples, such as distinguishing between white and black cigarettes, you can indeed leverage the embeddings from the YOLO-World model for downstream tasks. Here's a concise guide on how you can achieve this:

### Extracting Features for Downstream Tasks

1. **Feature Extraction**:
   You can extract features from the intermediate layers of the YOLO-World model. This involves modifying the model to output the desired feature maps. Based on your provided architecture, you can choose the appropriate layer from which to extract features.

2. **Using Extracted Features**:
   Once you have the feature maps, you can use them as input to a downstream classifier, such as a logistic regression model, similar to the approach you mentioned with CLIP.

Here’s a simplified example to guide you through the process:

```python
import torch
from ultralytics import YOLOWorld

# Load the YOLO-World model
model = YOLOWorld("yolov8s-worldv2.pt")

# Modify the forward method to extract intermediate features
class FeatureExtractor(torch.nn.Module):
    def __init__(self, model):
        super(FeatureExtractor, self).__init__()
        self.model = model

    def forward(self, x):
        features = []
        for name, module in self.model.model.named_children():
            x = module(x)
            if name in ['layer_name']:  # Replace 'layer_name' with the actual layer names you want to extract
                features.append(x)
        return features, x

# Initialize the feature extractor
feature_extractor = FeatureExtractor(model)

# Example input
input_image = torch.randn(1, 3, 640, 640)  # Replace with actual image tensor

# Extract features
features, output = feature_extractor(input_image)

# Use features for downstream tasks
# For example, train a logistic regression model on these features
```

### Fine-Grained Classification

For fine-grained classification, you can follow a similar approach to the CLIP example you provided. Extract the features using the modified YOLO-World model and then train a classifier on these features.

### Custom Prompts

Additionally, you can set custom prompts to guide the model's focus towards specific objects of interest without retraining:

```python
from ultralytics import YOLOWorld

# Initialize a YOLO-World model
model = YOLOWorld("yolov8s-world.pt")

# Define custom classes
model.set_classes(["white cigarette", "black cigarette"])

# Execute prediction for specified categories on an image
results = model.predict("path/to/image.jpg")

# Show results
results[0].show()
```

### Saving and Loading Custom Models

You can save the model with custom classes for future use:

```python
# Save the model with the defined offline vocabulary
model.save("custom_yolov8s.pt")

# Load your custom model
model = YOLOWorld("custom_yolov8s.pt")

# Run inference to detect your custom classes
results = model.predict("path/to/image.jpg")
results[0].show()
```

This approach allows you to tailor the model to your specific needs efficiently.

For more detailed information, you can refer to the [YOLO-World documentation](https://docs.ultralytics.com/models/yolo-world/).

I hope this helps! If you have any further questions, feel free to ask. 😊
